docs/combinatorics.rst

```diff
@@ -1,6 +1,210 @@
+leaves sum to :math:`n`.
+This maps very closely to the concept of integer partitions.
+Each tree shape of :math:`n` leaves can be represented by taking a
+nondecreasing integer partition of :math:`n` (elements of the partition
+are sorted in nondecreasing order) and recursively partitioning its
+elements. The order in which we select partitions of :math:`n` is
+determined by the efficient
+`rule_asc <http://jeromekelleher.net/generating-integer-partitions.html>`_
+algorithm for generating them.
+
+All tree shapes with four leaves, and the partitions that generate
+them, are:
+
+.. image:: _static/four_leaf_tree_shapes.png
+   :alt: All four-leaf tree shapes and their generating partitions
+
+Note that the middle column reflects all tree shapes of three leaves
+in the right subtree!
+
+`*` This excludes the partition [:math:`n`], since this would create a unary node
+and trees with unary nodes are inumerable (and potentially infinite).
+
+.. note::
+    Using nondecreasing integer partitions enforces a
+    *canonical orientation* on the tree shapes, where children under a node are
+    ordered by the number of leaves below them.
+    This is important because it prevents us from repeating trees that are
+    topologically the same but whose children are ordered differently.
+
+*********************
+Labelling Tree Shapes
+*********************
+
+Tree shapes are useful in and of themselves, but we can use the enumeration
+formulated above to go further and assign labels to the leaves of each shape.
+
+Say we are given a tree :math:`T` with :math:`n` leaves, whose left-most
+subtree, :math:`T_l`, has `k` leaves. For each of the :math:`n \choose k`
+ways to select labels to assign to :math:`T_l`, we produce a unique labelling
+of :math:`T`. This process of choosing labels is repeated for the other
+children of :math:`T` and then recursively for the subtrees.
+
+Looking back to the example from :ref:`sec_tree_ranks`, we can see
+the different unique ways to label a particular tree of three leaves.
+
+.. image:: _static/topology_1_0.svg
+    :width: 32%
+.. image:: _static/topology_1_1.svg
+    :width: 32%
+.. image:: _static/topology_1_2.svg
+    :width: 32%
+
+The order of the tree labellings is a direct result of the way in which
+combinations of labels are chosen. The implementation in tskit uses a
+standard lexicographic ordering to choose labels. See how the trees
+are sorted by the order in which the left leaf's label was chosen.
+
+.. note::
+    There is a caveat here regarding symmetry, similar to that of repeating
+    tree shapes. Symmetrical trees run the risk of creating redundant labellings
+    if all combinations of labels were exhausted. To prevent redundant labellings
+    we impose a *canonical labelling*. In the case of two symmetrical subtrees,
+    the left subtree must receive the minimum label from the label set. Notice
+    how this is the case in the right subtrees above.
```

Comment:
Gee, also, I think you had a good description of what the order on labellings is, before? It'd be nice to put that back in.